Role that does basic server health/sanity checks

jbzdak opened the pull request:
A generic sanity-checker tool, that can do following: 
- Check free space on given mounts. 
- Check whether some ports are open and accepting connections
- Run some external commands and watch their exit status

If any of the checks fail it will email root account. If all check pass it will optionally notify dead man's snitch.

antoviaque commented on defaults/main.yml:
```diff
@@ -0,0 +1,28 @@
+---
+# defaults file for sanity-check
+
+SANITY_CECK_DISK_SPACE:
```
Typo

jbzdak replied:
Addressed

antoviaque commented on defaults/main.yml:
```diff
@@ -0,0 +1,28 @@
+# If any of these return false
+SANITY_CHECK_COMMANDS:
+  - command: echo 1
+    description: "Echo broke, were doomed!"
```
"we are"

antoviaque commented on defaults/main.yml:
```diff
@@ -0,0 +1,28 @@
+
+SANITY_CHECK_SNITCH: null
+
+SANITY_CHECK_SUBJECT: 'Some errors in your server'
```
"Sanity check failed on `name.of.server`!" ?

jbzdak replied:
Addressed

antoviaque commented on templates/sanity-check.sh:
```diff
@@ -0,0 +1,48 @@
+report="{{ SANITY_CHECK_LOG_LOCATION }}"
+echo > ${report};
+
+{% for mount in SANITY_CECK_DISK_SPACE %}
```
Typo

antoviaque commented on defaults/main.yml:
```diff
@@ -0,0 +1,19 @@
+    # Mount point
+  - mount: /
+    # Treshold in mb
+    treshold_mb: 5000
```
Maybe worth using % instead of an actual number of MB, as this is going to be very dependent on the actual size of the drive. Maybe a default threshold of 80% of the hard drive space?

jbzdak replied:
This is doable. For me mb's are cleaner as if I can estimate how long it will take me to increase the drive size, and how much data is stored I can estimate safety marigin in mbs. But percents are fine too.

antoviaque replied:
Alright - so that still needs to be done, right?

jbzdak replied:
There is a new setting: `SANITY_CHECK_DISK_SPACE_PERCENTAGE` that works for all the fses under `/dev`, and I deleted `SANITY_CHECK_DISK_SPACE`. It might make sense to require that all fs have at least 20% free space at all times --- when I used mb's I needed to set per-partition limits, but in percents I could use global.

antoviaque commented on defaults/main.yml:
```diff
@@ -0,0 +1,19 @@
+
+SANITY_CHECK_DISK_SPACE:
+    # Mount point
+  - mount: /
```
Could there be a '*' value which would check all mounted drives? That would be a useful default here, to not have to remember updating this every time a partition changes or is added.

jbzdak replied:
In the current implementation no. But this can be revorked easily.

However we don't want to check all mounted filesystems, `df -h` shows the following: 

```
udev            1.9G   12K  1.9G   1% /dev
tmpfs           386M  372K  386M   1% /run
/dev/vda1        20G  3.2G   16G  17% /
none            4.0K     0  4.0K   0% /sys/fs/cgroup
none            5.0M     0  5.0M   0% /run/lock
none            1.9G     0  1.9G   0% /run/shm
none            100M     0  100M   0% /run/user
```

And I'm not sure what to do if `/run/lock` runs out of space. So maybe monitor just drives filesystems with `/dev`?

antoviaque replied:
Good point - limiting it to filesystems on /dev seem like a good idea, yes.

antoviaque commented on files/sanity_check.py:
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
```
This file doesn't have tests currently? I know we are not testing the ansible scripts themselves, but it might be worth testing this one

jbzdak replied:
Yes it does :) Not [100% coverage](https://github.com/open-craft/ansible-sanity-checker/blob/jbzdak/sanity-checker/tests/test_sanity_check.py) but covers all checks --- it would be rather bad if they failed :) If there is any configuration issue, this most probably would fail deployment, as sanity check is run at every deployment to check config.

antoviaque replied:
Ah, it's right below : )

antoviaque commented on files/sanity_check.py:
```diff
@@ -0,0 +1,111 @@
+import urllib2
+
+
+REPORT = []
```
Can we avoid using a global variable, and concatenate returned reports?

jbzdak replied:
Yes global state makes testing everything else harder. This is now addressed.

antoviaque commented on files/sanity_check.py:
```diff
@@ -0,0 +1,111 @@
+      process = subprocess.Popen(['mail', '-s', subject, recipient], stdin=subprocess.PIPE)
+      process.communicate(body)
+    except Exception as  error:
+        print(error)
```
Why is this exception caught and simply printed, instead of letting it propagate?

jbzdak replied:
No real reason (apart from me being tired :(). In the long run it has no consequence whatsoever, Becouse mail doesn't really catch much errors, I moved send_message to the last step of sanity check, so this is after snitch, and output from this command is ignored in cron. 

Nevertheless this is now fixed.

antoviaque commented on files/sanity_check.py:
```diff
@@ -0,0 +1,158 @@
+
+
+def test_global_free_percentage(default_free_space_percentage):
+    # TODO: it would be very good to add test for this one
```
+1

jbzdak replied:
Fixed in 837fff6ee68871d773cdd2e9c9994e660bb5b022 (PR #2)

antoviaque commented on files/sanity_check.py:
```diff
@@ -0,0 +1,158 @@
+
+    if report_is_ok(report):
+        if 'snitch' in json_data and json_data['snitch']:
+            if not check_heartbeat_url(json_data['snitch']):
```
Why are you using a function with such a specific name to get another URL than the heartbeat?

jbzdak replied:
I'll rename the function.

jbzdak replied:
Fixed in 837fff6ee68871d773cdd2e9c9994e660bb5b022 (PR #2)

antoviaque commented:
:+1: once the comments are addressed & the tests pass. If needed for the merge today, the changes can be done in a separate PR (but still part of the same task).

Commit:
Initial version of sanity-checker role

Squashed commit of the following:

commit 25b48016ab3f262d510d48cf3f70b890d8392b9b
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 19:47:07 2016 +0200

    Invert boolean condition

commit 7b5f6d3eb69d2a9c7ecbd25de3665163b56ddad3
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 17:58:20 2016 +0200

    Added messages to positive responses

commit dfeefa43f03b784b4d0a5e1120edc30ed3b318a5
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 17:46:12 2016 +0200

    Added better report formatting

commit 04b32cde1e32bf794a71a47e8d756040d27f72b1
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 15:51:43 2016 +0200

    Updated readme

commit 29cb8a68daf9bbf93e86e1a69cf7dcf6b3277a67
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 15:31:19 2016 +0200

    Updated readme

commit 025e36936521ec07d3ec80abd219c03c9cbb3234
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Sun Apr 24 22:37:58 2016 +0200

    Another small fix

commit 5413b1f0fa87672f575652fcb0ce3ede7489505f
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Sun Apr 24 21:59:16 2016 +0200

    Small update

commit 071d6bbe005072cec65122978647187d8349c5ff
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Sun Apr 24 12:08:50 2016 +0200

    Fix sanity checker

commit 8c7a536bc88867fd75f7081b7a806ae811a86fd3
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Thu Apr 21 22:19:53 2016 +0200

    Small fix in sanity checker

commit b32d479643f0872da976413c8f2766d31dcd48bc
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Thu Apr 21 13:37:49 2016 +0200

    Fix for sanity check

commit 4f4fd49f3afbfdeacb53a684ac9c88d253afe6db
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Tue Apr 19 23:01:11 2016 +0200

    PR fixes

commit ad21c2adc15fe0ca0813b6bdca9d4a4a8c5748b5
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Tue Apr 19 22:57:25 2016 +0200

    PR review fixes

commit 0892a7b3d00dfffbf8b45908bde77de84ea8b3af
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 18 13:00:20 2016 +0200

    Role that does basic server health/sanity checks

jbzdak commented:
Squashed and merging, will adress concerns in a separate PR
